File: solutions/Arrays & Hashing/36_valid-sudoku.py
from typing import List

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        
        for i in range(9):
            row = set()
            for j in range(9):
                if board[i][j] != ".":
                    if board[i][j] in row:
                        return False
                    row.add(board[i][j])

        for j in range(9):
            col = set()
            for i in range(9):
                if board[i][j] != ".":
                    if board[i][j] in col:
                        return False
                    col.add(board[i][j])

        start = [(0,0), (0,3), (0,6), (3,0), (3,3), (3,6), (6,0), (6,3), (6,6)]

        for x, y in start:
            box = set()
            for i in range(x, x+3):
                for j in range(y, y+3):
                    if board[i][j] != ".":
                        if board[i][j] in box:
                            return False
                        box.add(board[i][j])
        return True
    
# Time Complexity: O(1) - board size is fixed
# Space Complexity: O(1) - board size is fixed


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:

        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        box = [set() for _ in range(9)]
        
        for i in range(9):
            for j in range(9):
                x = board[i][j]
                if x != ".":
                    if x in rows[i] or x in cols[j] or x in box[(i//3)*3 + j//3]:
                        return False

                    rows[i].add(x)
                    cols[j].add(x)
                    box[(i//3) *3 + j//3].add(x)

        return True
# [One Pass Solution]
# Time Complexity: O(1) - board size is fixed
# Space Complexity: O(1) - board size is fixed

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:

        rows = [0] * 9
        cols = [0] * 9
        box = [0] * 9
        
        for i in range(9):
            for j in range(9):
                if board[i][j] != ".":
                    val = 1 << (int(board[i][j]) - 1)
                    if val & rows[i] or val & cols[j] or val & box[(i//3)*3 + j//3]:
                        return False

                    rows[i] |= val
                    cols[j] |= val
                    box[(i//3) *3 + j//3] |= val

        return True

# [BitMask Solution]
# Time Complexity: O(1) - board size is fixed
# Space Complexity: O(1) - board size is fixed

File: C++/Arrays & Hashing/36_valid-sudoku.cpp
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int n = 9;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] != '.' && !check(board, i, j)) {
                    return false;
                }
            }
        }
        return true;
    }
    bool check(vector<vector<char>>& board, int x, int y) {
        int n = 9;
        char val = board[x][y];
        int x1 = (x / 3) * 3;
        int y1 = (y / 3) * 3;

        for (int i = x1; i < x1+3; ++i) {
            if (i == x) continue;
            for (int j = y1; j < y1+3; ++j) {
                if (j != y && board[i][j] == val) return false;
            }
        }

        int j = y;
        for (int i = 0; i < n; ++i) {
            if (i != x && board[i][j] == val) return false;
        }

        int i = x;
        for (int j = 0; j < n; ++j) {
            if (j != y && board[i][j] == val) return false;
        }

        return true;
    }
};


class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int n = 9;
        vector<unordered_set<char>> row(9);
        vector<unordered_set<char>> col(9);
        vector<unordered_set<char>> box(9);

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] != '.') {
                    char val = board[i][j];
                    int b = ((i / 3) * 3) + (j / 3);

                    if (row[i].contains(val) || col[j].contains(val) || box[b].contains(val)) {
                        return false;
                    }
                    row[i].insert(val);
                    col[j].insert(val);
                    box[b].insert(val);
                }
            }
        }
        return true;
    }
};


class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int n = 9;
        vector<int> row(9);
        vector<int> col(9);
        vector<int> box(9);

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] != '.') {
                    int b = ((i / 3) * 3) + (j / 3);
                    int val = board[i][j] - '0';

                    if (row[i] & 1 << val || col[j] & 1 << val || box[b] & 1 << val) {
                        return false;
                    }
                    row[i] |= 1 << val;
                    col[j] |= 1 << val;
                    box[b] |= 1 << val;
                }
            }
        }
        return true;
    }
};